change default aws linux instance type to t3.micro, reduce default template TTL

johnstcn opened the pull request.

johnstcn committed:
change default aws linux instance type to t3.micro, reduce default template TTL

johnstcn commented on cli/templatecreate.go:
```diff
@@ -127,7 +127,7 @@ func templateCreate() *cobra.Command {
 	cmd.Flags().StringVarP(&provisioner, "test.provisioner", "", "terraform", "Customize the provisioner backend")
 	cmd.Flags().StringVarP(&parameterFile, "parameter-file", "", "", "Specify a file path with parameter values.")
-	cmd.Flags().DurationVarP(&maxTTL, "max-ttl", "", 168*time.Hour, "Specify a maximum TTL for worksapces created from this template.")
+	cmd.Flags().DurationVarP(&maxTTL, "max-ttl", "", 8*time.Hour, "Specify a maximum TTL for workspaces created from this template.")
```
```suggestion
	cmd.Flags().DurationVarP(&maxTTL, "max-ttl", "", 2*time.Hour, "Specify a maximum TTL for workspaces created from this template.")
```
```suggestion
	cmd.Flags().DurationVarP(&maxTTL, "max-ttl", "", 8*time.Hour, "Specify a maximum TTL for workspaces created from this template.")
```

johnstcn committed:
Update cli/templatecreate.go

kylecarbs commented on cli/templatecreate.go:
```diff
@@ -127,7 +127,7 @@ func templateCreate() *cobra.Command {
 	cmd.Flags().StringVarP(&provisioner, "test.provisioner", "", "terraform", "Customize the provisioner backend")
 	cmd.Flags().StringVarP(&parameterFile, "parameter-file", "", "", "Specify a file path with parameter values.")
-	cmd.Flags().DurationVarP(&maxTTL, "max-ttl", "", 168*time.Hour, "Specify a maximum TTL for worksapces created from this template.")
+	cmd.Flags().DurationVarP(&maxTTL, "max-ttl", "", 2*time.Hour, "Specify a maximum TTL for workspaces created from this template.")
```
Do we display the max TTL in the UI anywhere? I'm concerned about setting this so low without a clear indicator to the user.

johnstcn replied:
That's fair!

spikecurtis commented on examples/templates/aws-linux/main.tf:
```diff
@@ -130,7 +130,7 @@ EOT
   ami               = data.aws_ami.ubuntu.id
   availability_zone = "${var.region}a"
-  instance_type     = "t3.xlarge"
+  instance_type     = "t3.micro"
```
This is going to give people a poor experience with cloud workspaces.  I understand the desire to keep our own spend low when these things get orphaned, but I think we need to do this in a way that doesn't cripple end users experience.

We want people to have a good experience with cloud development and get hooked.  A micro is less powerful than my phone.

johnstcn replied:
This bring it in line with the AWS windows template. Should be instead update the AWS windows template to match?
Alternatively, we could instead force the user to choose an instance type when creating a workspace via the template.
What do you think?

spikecurtis replied:
I think modeling it as a variable is a good idea for our example templates.  Give them a reasonable range but keep the default something powerful.  Hopefully our own devs who are just testing, not dogfooding, will choose something small.

johnstcn committed:
- make default template max TTL 24 hours (still less than 168)
- make default workspace autostop 2 hours unless specified otherwise
- add instance type selector to aws templates

mafredri commented on examples/templates/aws-linux/main.tf:
```diff
@@ -36,6 +36,22 @@ variable "region" {
 
+variable "instance_type" {
+  description = "What instance type should your workspace use?"
+  default     = "2C/1G: t3.micro"
```
I've had a few instances where defining the `default` skipped the prompt. I'm guessing this is not the case here? (Just checking.)

johnstcn replied:
I don't recall it being the case when I tested it!

mafredri reviewed (approved):
Default TTL of 2h sounds a bit short, but I understand why we're making the change, so all good 👍🏻.

johnstcn committed:
Merge branch 'main' into cj/template-ttl-default

spikecurtis commented on examples/templates/aws-linux/main.tf:
```diff
@@ -36,6 +36,22 @@ variable "region" {
+  validation {
+    condition = contains([
+      "2C/1G: t3.micro",
+      "2C/2G: t3.small",
```
It wasn't immediately clear what "C" and "G" meant in this context.  I eventually realized it was vCPU and GB RAM, but I think 2 vCPU /  2 GB RAM is clearer and it's not like we're hurting for space in this context, nor do the users have to type these values in our UX.

Unfortunately the vCPU number is a little misleading for T3 instances. Micro-large have the same number of vCPUs, and reading this you might conclude that they have the same CPU performance.  They do not.  Amazon throttles the CPUs except for short bursts and the amount of throttling is different for the different instance sizes.  I'm not really sure what to suggest except to drop the stats and just include the instance type names.  Up to you.

johnstcn replied:
I can drop the stats, I only put them in there so the instance types would show up in the right order :D

mafredri replied:
I guess you could do numbered list, `1. t3.micro`, `2. t3.small` if you want to maintain order? 😄

johnstcn committed:
remove prefixes from instance types
